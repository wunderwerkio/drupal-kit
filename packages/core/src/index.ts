export * from "./Drupalkit";
export * from "./types";
