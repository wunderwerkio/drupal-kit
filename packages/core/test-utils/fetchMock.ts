import type { FetchMock } from "jest-fetch-mock";

import type { Drupalkit } from "../src";

export const mockResponse = (
  fm: FetchMock,
  dk: Drupalkit,
  options: {
    url: string | RegExp;
    status?: number;
    payloadFixture?: object;
    locale?: string;
    contentType?: string;
  },
) => {
  const url =
    typeof options.url === "string"
      ? dk.buildUrl(options.url, {
          localeOverride: options.locale,
        })
      : options.url;

  // Mock next request for url.
  fm.mockOnceIf(url, async (req) => {
    const payload = options.payloadFixture;

    return {
      body: payload ? JSON.stringify(payload) : "{}",
      headers: {
        "content-type": options.contentType ?? "application/json",
      },
      url: req.url,
      status: options.status ?? 200,
    };
  });
};

export const mockNetworkError = (fm: FetchMock) => {
  fm.mockRejectOnce(new Error("Network Error"));
};
