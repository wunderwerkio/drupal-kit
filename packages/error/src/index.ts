import * as DrupalkitTypes from "@drupalkit/types";

import { RequestErrorOptions as DrupalkitErrorOptions } from "./types";

/**
 * Custom error class to help with error handling.
 */
export class DrupalkitError extends Error {
  /**
   * Error name.
   */
  name: "HttpError";

  /**
   * HTTP status code.
   */
  status: number;

  /**
   * Request options that lead to the error.
   */
  request: DrupalkitTypes.RequestOptions;

  /**
   * Response object if a response was received.
   */
  response?: DrupalkitTypes.DrupalkitResponse<unknown>;

  /**
   * Construct a new DrupalkitError.
   *
   * @param message - Error message.
   * @param statusCode - HTTP status code.
   * @param options - Request / Response that lead to the error.
   */
  constructor(
    message: string,
    statusCode: number,
    options: DrupalkitErrorOptions,
  ) {
    super(message);

    // Maintains proper stack trace (only available on V8)
    // @ts-ignore
    if (Error.captureStackTrace) {
      // @ts-ignore
      Error.captureStackTrace(this, this.constructor);
    }

    this.name = "HttpError";
    this.status = statusCode;

    if (options.response) {
      this.response = options.response;
    }

    // redact request credentials without mutating original request options
    const requestCopy = Object.assign({}, options.request);
    if (options.request.headers.authorization) {
      requestCopy.headers = Object.assign({}, options.request.headers, {
        authorization: options.request.headers.authorization.replace(
          / .*$/,
          " [REDACTED]",
        ),
      });
    }

    this.request = requestCopy;
  }
}
