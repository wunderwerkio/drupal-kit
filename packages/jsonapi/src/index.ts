export * from "./resources";
export * from "./DrupalkitJsonApi";
