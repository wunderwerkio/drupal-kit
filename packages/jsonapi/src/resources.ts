import { DrupalJsonApiParamsInterface } from "drupal-jsonapi-params";
import {
  LinkObject,
  Relationship,
  ResourceObject,
  Response,
} from "ts-json-api";

export interface JsonApiIndex extends Response<[]> {
  links: {
    [type: string]: LinkObject;
  };
}

type ResourceCreatePayload<R extends ResourceObject> = {
  data: Pick<R, "type" | "attributes" | "relationships"> &
    Partial<Pick<R, "id">>;
};

type ResourceUpdatePayload<R extends ResourceObject> = {
  data: Pick<R, "id" | "type"> & {
    attributes?: Partial<R["attributes"]>;
    relationships?: Partial<R["relationships"]>;
  };
};

type ReadParameters = {
  queryParams?: DrupalJsonApiParamsInterface;
};

export type ReadSingleParameters = ReadParameters & {
  uuid: string;
};

export type ReadManyParameters = ReadParameters;

export type CreateParameters<Resource extends ResourceObject> = {
  payload: ResourceCreatePayload<Resource>;
};

export type UpdateParameters<Resource extends ResourceObject> = {
  uuid: string;
  payload: ResourceUpdatePayload<Resource>;
};

export type DeleteParameters = {
  uuid: string;
};

export type ToParameters<
  Operation,
  Resource extends ResourceObject,
> = "readSingle" extends Operation
  ? ReadSingleParameters
  : "readMany" extends Operation
  ? ReadManyParameters
  : "create" extends Operation
  ? CreateParameters<Resource>
  : "update" extends Operation
  ? UpdateParameters<Resource>
  : "delete" extends Operation
  ? DeleteParameters
  : never;



























interface FileResource extends ResourceObject {
  type: "file--file";
  attributes: {
    langcode: string;
    filename: string;
    uri: {
      value: string;
      url: string;
    };
    filemime: string;
    filesize: number;
    status: boolean;
    created: string;
    changed: string;
  };
  relationships: {
    uid: Relationship<UserResource>;
  };
}

interface UserResource extends ResourceObject {
  type: "user--user";
  attributes: {
    field_firstname: string;
    field_lastname: string;
    created: string;
    changed: string;
  };
}


export interface JsonApiResources {
  "file--file": {
    resource: FileResource;
    operations: "readSingle" | "readMany" | "create" | "update" | "delete";
  };
}
