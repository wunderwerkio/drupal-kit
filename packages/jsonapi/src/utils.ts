import { RequestOptions } from "@drupal-kit/types";
import { getHeader } from "@drupal-kit/utils";

/**
 * Checks if request is a JSON:API request.
 *
 * @param req - The request options.
 */
export const isJsonApiRequest = (req: RequestOptions) => {
  const contentType = getHeader("content-type", req);

  return contentType === "application/vnd.api+json";
};
