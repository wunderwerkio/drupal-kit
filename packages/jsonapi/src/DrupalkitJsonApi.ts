import { ResourceObject, Response } from "ts-json-api";
import { Ok } from "ts-results";
import { Drupalkit, DrupalkitOptions, Query } from "@drupalkit/core";

import {
  JsonApiIndex,
  JsonApiResources,
  ReadManyParameters,
  ReadSingleParameters,
  ToParameters,
} from "./resources.js";

declare module "@drupalkit/core" {
  interface DrupalkitOptions {
    jsonApiPrefix?: string;
  }
}

/**
 * Drupalkit plugin that provides integration with the Drupal JSON:API.
 *
 * @param drupalkit - The Drupalkit instance.
 * @param drupalkitOptions - The options for the Drupalkit instance.
 */
export const DrupalkitJsonApi = (
  drupalkit: Drupalkit,
  drupalkitOptions: DrupalkitOptions,
) => {
  const defaultHeaders: HeadersInit = {
    Accept: "application/vnd.api+json",
    "Content-Type": "application/vnd.api+json",
  };

  /**
   * Retrieves the JSON:API index.
   *
   * @returns A result object containing the JSON:API index or an error.
   */
  const getIndex = async () => {
    const url = buildJsonApiUrl("");

    const response = await drupalkit.request<JsonApiIndex>(url, {
      method: "GET",
      headers: {
        ...defaultHeaders,
      },
    });

    if (response.err) {
      return response;
    }

    return Ok(response.val.data);
  };

  /**
   * Retrieves a single JSON:API resource object.
   *
   * @param type - The type of resource object to retrieve.
   * @param parameters - The parameters to use for the query.
   * @param options - Optional settings to override locale and default locale.
   * @param options.localeOverride - An optional override for the locale.
   * @param options.defaultLocaleOverride - An optional override for the default locale.
   * @returns A result object containing the resource object or an error.
   */
  const getResource = async <R extends ResourceObject>(
    type: R["type"],
    parameters: ReadSingleParameters,
    options?: {
      localeOverride?: string;
      defaultLocaleOverride?: string;
    },
  ) => {
    const path = type.replace("--", "/") + "/" + parameters.uuid;

    const url = buildJsonApiUrl(path, {
      ...options,
      query: parameters.queryParams?.getQueryObject(),
    });

    const result = await drupalkit.request<Response<R>>(url, {
      method: "GET",
      headers: defaultHeaders,
    });

    if (result.err) {
      return result;
    }

    return Ok(result.val.data);
  };

  /**
   * Retrieves multiple JSON:API resource objects.
   *
   * @param type - The type of resource object to retrieve.
   * @param parameters - The parameters to use for the query.
   * @param options - Optional settings to override locale and default locale.
   * @returns A result object containing the resource object or an error.
   */
  const getResourceCollection = async <R extends ResourceObject>(
    type: R["type"],
    parameters: ReadManyParameters,
    options?: {
      localeOverride?: string;
      defaultLocaleOverride?: string;
    },
  ) => {
    const path = type.replace("--", "/");

    const url = buildJsonApiUrl(path, {
      ...options,
      query: parameters.queryParams?.getQueryObject(),
    });

    const result = await drupalkit.request<Response<R[]>>(url, {
      method: "GET",
      headers: defaultHeaders,
    });

    if (result.err) {
      return result;
    }

    return Ok(result.val.data);
  };

  /**
   * Constructs a JSON API URL for use with Drupal.
   *
   * @param path - The path to the JSON API endpoint.
   * @param options - An optional object containing additional options.
   * @param options.localeOverride - An optional override for the locale.
   * @param options.defaultLocaleOverride - An optional override for the default locale.
   * @returns The constructed URL as a string.
   */
  const buildJsonApiUrl = (
    path: string,
    options?: {
      localeOverride?: string;
      defaultLocaleOverride?: string;
      query?: Query;
    },
  ) => {
    const prefix = drupalkitOptions.jsonApiPrefix ?? "jsonapi";

    return drupalkit.buildUrl(path, {
      ...options,
      customPrefix: prefix,
    });
  };

  /**
   * Extend the Drupalkit instance.
   */
  return {
    jsonApi: {
      buildJsonApiUrl,
      getIndex,
      async resource<
        Type extends keyof JsonApiResources,
        Resource extends JsonApiResources[Type]["resource"],
        Operation extends JsonApiResources[Type]["operations"],
        Params extends ToParameters<Operation, Resource>,
      >(
        type: Type,
        operation: Operation,
        parameters: Params,
        options?: {
          localeOverride?: string;
          defaultLocaleOverride?: string;
        },
      ) {
        switch (operation) {
          case "readSingle":
            return getResource(
              type,
              parameters as ReadSingleParameters,
              options,
            );
          case "readMany":
            return getResourceCollection(
              type,
              parameters as ReadManyParameters,
              options,
            );
        }
      },
    },
  };
};
