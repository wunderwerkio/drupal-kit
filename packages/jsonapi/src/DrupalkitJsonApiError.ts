import { DrupalkitError, DrupalkitErrorOptions } from "@drupal-kit/error";

type JsonApiError = {
  title: string;
  status: string;
  detail: string;
  source?: {
    pointer?: string;
  };
};

/**
 * Enhanced DrupalkitError with JSON:API specific data.
 */
export class DrupalkitJsonApiError extends DrupalkitError {
  /**
   * Array of JSON:API errors.
   */
  readonly errors: JsonApiError[];

  /**
   * @inheritdoc
   */
  constructor(
    message: string,
    statusCode: number,
    errors: JsonApiError[],
    options: DrupalkitErrorOptions,
  ) {
    super(message, statusCode, options);

    this.errors = errors;
  }

  /**
   * Checks if this error instance contains validation errors.
   */
  public hasValidationErrors() {
    return this.errors.some((error) => error.status == "422");
  }

  /**
   * Get an array of all fields with invalid values.
   */
  public getInvalidFields() {
    return this.getErrorsByStatus(422)
      .map((error) => {
        const pointer = error?.source?.pointer;

        if (!pointer) {
          return null;
        }

        return pointer.split("/").pop();
      })
      .filter((field) => field !== null);
  }

  /**
   * Get all errors with the given status code.
   *
   * @param statusCode - The status code to filter by.
   */
  public getErrorsByStatus(statusCode: number) {
    return this.errors.filter(
      (error) => error.status.toString() == statusCode.toString(),
    );
  }

  /**
   * Create new error instance from a DrupalkitError.
   *
   * @param error - The DrupalkitError to create an instance from.
   */
  public static fromDrupalkitError(error: DrupalkitError) {
    const errors = error.response
      ? this.extractJsonApiErrorsFromResponse(error.response.data)
      : [];

    return new DrupalkitJsonApiError(error.message, error.statusCode, errors, {
      request: error.request,
      response: error.response,
    });
  }

  /**
   * Extract JSON API errors from response data.
   *
   * @param data - The response data to extract errors from.
   */
  public static extractJsonApiErrorsFromResponse(
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    data: any,
  ) {
    if (!data?.errors) {
      return [];
    }

    return data.errors as JsonApiError[];
  }
}
