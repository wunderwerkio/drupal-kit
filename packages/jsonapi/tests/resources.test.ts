import { DrupalJsonApiParams } from "drupal-jsonapi-params";
import fetchMock, { enableFetchMocks } from "jest-fetch-mock";
import { Drupalkit } from "@drupalkit/core";
import { DrupalkitError } from "@drupalkit/error";

import { DrupalkitJsonApi } from "../src/index";
import JsonApiIndex from "./fixtures/jsonapi_index.json";
import JsonApiIndexError from "./fixtures/jsonapi_index_error.json";

enableFetchMocks();

describe("resources", () => {
  const BASE_URL = "https://my-drupal.com";

  const createDrupalkit = ({
    baseUrl = BASE_URL,
  }: { baseUrl?: string } = {}) => {
    const EnhancedDrupalkit = Drupalkit.plugin(DrupalkitJsonApi);

    return new EnhancedDrupalkit({
      baseUrl,
      locale: "de",
      defaultLocale: "de",
    });
  };

  it("should build json api url", () => {
    const drupalkit = createDrupalkit();
    const queryParams = new DrupalJsonApiParams();
    queryParams
      .addGroup("group")
      .addFilter("status", "1", "=", "test")
      .addSort("name", "ASC")
      .addInclude(["one", "two"])
      .addCustomParam({
        revision: "5",
      });

    // Simple url.
    let url = drupalkit.jsonApi.buildJsonApiUrl("node/article");
    expect(url).toMatchSnapshot("simple");

    // With query.
    url = drupalkit.jsonApi.buildJsonApiUrl("node/article", {
      query: queryParams.getQueryObject(),
    });
    expect(url).toMatchSnapshot("with-query");

    // With locale.
    url = drupalkit.jsonApi.buildJsonApiUrl("node/article", {
      localeOverride: "es",
    });
    expect(url).toMatchSnapshot("with-locale");

    // With base locale.
    url = drupalkit.jsonApi.buildJsonApiUrl("node/article", {
      localeOverride: "es",
      defaultLocaleOverride: "es",
    });
    expect(url).toMatchSnapshot("with-base-locale");

    // With locale and query.
    url = drupalkit.jsonApi.buildJsonApiUrl("node/article", {
      localeOverride: "es",
      query: queryParams.getQueryObject(),
    });
    expect(url).toMatchSnapshot("with-locale-and-query");
  });

  it("should get jsonapi index", async () => {
    const drupalkit = createDrupalkit();

    fetchMock.mockResponseOnce(() =>
      Promise.resolve({
        body: JSON.stringify(JsonApiIndex),
        headers: {
          "content-type": "application/vnd.api+json",
        },
      }),
    );

    let index = await drupalkit.jsonApi.getIndex();

    expect(index.ok).toBeTruthy();

    if (index.ok) {
      expect(JSON.stringify(index.val)).toMatchSnapshot("jsonapi-index");

      expect(index.val).toHaveProperty("jsonapi");
      expect(index.val).toHaveProperty("data");
      expect(index.val).toHaveProperty("links");
    }

    // With error
    fetchMock.mockResponseOnce(() =>
      Promise.resolve({
        body: JSON.stringify(JsonApiIndexError),
        status: 500,
        headers: {
          "content-type": "application/vnd.api+json",
        },
      }),
    );

    index = await drupalkit.jsonApi.getIndex();

    expect(index.err).toBeTruthy();

    if (index.err) {
      expect(index.val).toBeInstanceOf(DrupalkitError);
    }
  });
});
