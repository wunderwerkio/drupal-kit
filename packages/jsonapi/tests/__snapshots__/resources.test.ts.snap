// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resources should build json api url: simple 1`] = `"https://my-drupal.com/jsonapi/node/article"`;

exports[`resources should build json api url: with-base-locale 1`] = `"https://my-drupal.com/jsonapi/node/article"`;

exports[`resources should build json api url: with-locale 1`] = `"https://my-drupal.com/es/jsonapi/node/article"`;

exports[`resources should build json api url: with-locale-and-query 1`] = `"https://my-drupal.com/es/jsonapi/node/article?filter%5Bgroup%5D%5Bgroup%5D%5Bconjunction%5D=OR&filter%5Bstatus%5D%5Bcondition%5D%5Bpath%5D=status&filter%5Bstatus%5D%5Bcondition%5D%5Bvalue%5D=1&filter%5Bstatus%5D%5Bcondition%5D%5BmemberOf%5D=test&include=one%2Ctwo&sort=name&revision=5"`;

exports[`resources should build json api url: with-query 1`] = `"https://my-drupal.com/jsonapi/node/article?filter%5Bgroup%5D%5Bgroup%5D%5Bconjunction%5D=OR&filter%5Bstatus%5D%5Bcondition%5D%5Bpath%5D=status&filter%5Bstatus%5D%5Bcondition%5D%5Bvalue%5D=1&filter%5Bstatus%5D%5Bcondition%5D%5BmemberOf%5D=test&include=one%2Ctwo&sort=name&revision=5"`;

exports[`resources should get jsonapi index: jsonapi-index 1`] = `"{"jsonapi":{"version":"1.0","meta":{"links":{"self":{"href":"http://jsonapi.org/format/1.0/"}}}},"data":[],"links":{"file--file":{"href":"https://my-drupal.com/jsonapi/file/file"},"media--image":{"href":"https://my-drupal.com/jsonapi/media/image"},"node--article":{"href":"https://my-drupal.com/jsonapi/node/article"},"node--page":{"href":"https://my-drupal.com/jsonapi/node/page"},"self":{"href":"https://my-drupal.com/jsonapi"},"user--user":{"href":"https://my-drupal.com/jsonapi/user/user"}}}"`;
